docs/proposals/proposal-audit-log.md

```diff
@@ -0,0 +1,220 @@
+- `updateTargetSchemaComposition`
+
+### SchemaPublisher
+
+- `delete`
+
+### SchemaManager
+
+- `updateSchemaVersionStatus`
+- `createVersion`
+- `updateBaseSchema`
+- `disableExternalSchemaComposition`
+- `enableExternalSchemaComposition`
+- `updateNativeSchemaComposition`
+- `updateRegistryModel`
+- `approveFailedSchemaCheck`
+
+### BillingProvider
+
+- `upgradeToPro`
+- `syncOrganization`
+- `downgradeToHobby`
+
+### CdnProvider
+
+- `createCDNAccessToken`
+- `deleteCDNAccessToken`
+
+### CollectionProvider
+
+- `createCollection`
+- `deleteCollection`
+- `createOperation`
+- `updateOperation`
+- `updateCollection`
+- `deleteOperation`
+
+### GitHubIntegrationManager
+
+- `register`
+- `unregister`
+- `enableProjectNameInGithubCheck`
+
+### SlackIntegrationManager
+
+- `register`
+- `unregister`
+
+### OIDCIntegrationsProvider
+
+- `createOIDCIntegrationForOrganization`
+- `updateOIDCIntegration`
+- `deleteOIDCIntegration`
+
+### Contracts
+
+- `createContract`
+- `disableContract`
+
+### AlertsManager
+
+- `addChannel`
+- `deleteChannels`
+- `addAlert`
+- `deleteAlerts`
+
+## Storage
+
+- updateUser
+- updateOrganizationName
+- updateOrganizationPlan
+- updateOrganizationRateLimits
+- createOrganizationInvitation
+- deleteOrganizationInvitationByEmail
+- createOrganizationTransferRequest
+- answerOrganizationTransferRequest
+- addOrganizationMemberViaInvitationCode
+- deleteOrganizationMember
+- updateOrganizationMemberAccess
+- assignOrganizationMemberRole
+- assignOrganizationMemberRoleToMany
+- deleteOrganizationMemberRole
+- updateProjectRegistryModel
+- createVersion
+- updateVersionStatus
+- createActivity
+- addSlackIntegration
+- deleteSlackIntegration
+- addGitHubIntegration
+- deleteGitHubIntegration
+- setSchemaPolicyForOrganization
+- setSchemaPolicyForProject
+- createDocumentCollection
+- deleteDocumentCollection
+- updateDocumentCollection
+- createDocumentCollectionDocument
+- deleteDocumentCollectionDocument
+- updateDocumentCollectionDocument
+- createSchemaCheck
+
+
+## Implementation
+
+Proposed implementation is to use a single Clickhouse table to store all events. The table will have
+the following columns:
+
+
+```sql
+CREATE TABLE audit_log (
+  event_time DateTime DEFAULT now(),
+  user_id UUID,
+  organization_id UUID,
+  project_id UUID,
+  project_name STRING,
+  target_id UUID,
+  target_name STRING,
+  schema_version_id UUID,
+  event_kind STRING,
+  event_action STRING,
+  event_details JSON
+) ENGINE = MergeTree ()
+ORDER BY event_time
+TTL timestamp + INTERVAL 3 MONTH;
+```
+Data in clickhouse would be append only. We would never manually delete them, we would rely on TTL removing old rows after X months
+Of course we could make the interval configurable if necessary.
+
+Our log function will be a simple function that inserts a row into the table:
+
+```ts
+import { ClickHouse } from 'clickhouse';
+
+const clickhouse = new ClickHouse(/* ... */)
+
+type AuditLogEvent = {
+  userId: string;
```

Comment:
not every event can be pointed to a user, access tokens are currently not linked to any user account AFAIK.